README.rst
====================================
Text-Aware Process Prediction Model
====================================

This package provides a process prediction model,
which is able to take textual data into account.

This work is part of my master's thesis:
https://github.com/davidgeorgi/thesis/blob/master/thesis/thesis.pdf


HISTORY.rst
=======
History
=======

0.1 (2020-09-22)
------------------

First release


tapp/log_encoder.py
import numpy as np
from abc import ABC, abstractmethod


class Encoder(ABC):

    def __init__(self):
        super().__init__()

    @abstractmethod
    def fit(self, docs):
        pass

    @abstractmethod
    def transform(self, docs):
        pass


class LogEncoder(Encoder):
    
    def __init__(self, text_encoder=None, advanced_time_attributes=True):
        self.text_encoder = text_encoder
        self.activities = []
        self.data_attributes = []
        self.text_attribute = None
        self.categorical_attributes = []
        self.categorical_attributes_values = []
        self.numerical_attributes = []
        self.numerical_divisor = []
        self.event_dim = 0
        self.feature_dim = 0
        self.advanced_time_attributes = advanced_time_attributes
        self.time_scaling_divisor = [1, 1, 1, 1, 1, 1]
        self.process_start_time = 0
        super().__init__()

    def fit(self, log, activities=None, data_attributes=None, text_attribute=None):
        # Fit encoder to log
        self.activities = activities
        self.data_attributes = data_attributes
        self.text_attribute = text_attribute
        self.categorical_attributes = list(filter(lambda attribute: not _is_numerical_attribute(log, attribute), self.data_attributes))
        self.categorical_attributes_values = [_get_event_labels(log, attribute) for attribute in self.categorical_attributes]
        self.numerical_attributes = list(filter(lambda attribute: _is_numerical_attribute(log, attribute), self.data_attributes))
        self.numerical_divisor = [np.max([event[attribute].timestamp() for case in log for event in case]) for attribute in self.numerical_attributes]
        self.process_start_time = np.min([event["time:timestamp"].timestamp() for case in log for event in case])

        # Scaling divisors for time related features to achieve values between 0 and 1
        cycle_time_max = np.max(
            [case[-1]["time:timestamp"].timestamp() - case[0]["time:timestamp"].timestamp() for case in log])
        time_between_events_max = np.max(
            [event["time:timestamp"].timestamp() - case[event_index - 1]["time:timestamp"].timestamp() for case in
             log for event_index, event in enumerate(case) if event_index > 0])
        log_time = np.max([event["time:timestamp"].timestamp() for case in log for event in case]) - self.process_start_time
        self.time_scaling_divisor = [cycle_time_max, time_between_events_max, 86400, 604800, 31536000, log_time]

        # Event dimension: Maximum number of events in a case
        self.event_dim = _get_max_case_length(log)

        # Feature dimension: Encoding size of an event
        activity_encoding_length = len(self.activities)
        time_encoding_length = 6 if self.advanced_time_attributes else 2
        categorical_attributes_encoding_length = sum([len(values) for values in self.categorical_attributes_values])
        numerical_attributes_encoding_length = len(self.numerical_attributes)
        text_encoding_length = self.text_encoder.encoding_length if self.text_encoder is not None and self.text_attribute is not None else 0
        self.feature_dim = self.feature_dim = activity_encoding_length + time_encoding_length + categorical_attributes_encoding_length + numerical_attributes_encoding_length + text_encoding_length

        # Train text encoder
        if self.text_encoder is not None and self.text_attribute is not None:
            docs = [event[self.text_attribute] for case in log for event in case if self.text_attribute in event]
            self.text_encoder.fit(docs)

    def transform(self, log, for_training=True):
        case_dim = np.sum([len(case) for case in log]) if for_training else len(log)

        # Prepare input and output vectors/matrices
        x = np.zeros((case_dim, self.event_dim, self.feature_dim))
        if for_training:
            y_next_act = np.zeros((case_dim, len(self.activities) + 1))
            y_final_act = np.zeros((case_dim, len(self.activities)))
            y_next_time = np.zeros(case_dim)
            y_final_time = np.zeros(case_dim)

        # Encode traces and prefix traces
        trace_dim_index = 0
        for case in log:
            case_start_time = case[0]["time:timestamp"].timestamp()
            # For training: Encode all prefixes. For predicting: Encode given prefix only
            prefix_lengths = range(1, len(case) + 1) if for_training else range(len(case), len(case) + 1)
            for prefix_length in prefix_lengths:
                # Encode the (prefix-)trace
                previous_event_time = case_start_time
                # Post padding of event sequences
                padding = self.event_dim - prefix_length
                for event_index, event in enumerate(case):

                    if event_index <= prefix_length - 1:
                        # Encode activity
                        if event["concept:name"] in self.activities:
                            x[trace_dim_index][padding+event_index][self.activities.index(event["concept:name"])] = 1
                        offset = len(self.activities)

                        # Encode time attributes
                        event_time = event["time:timestamp"]
                        # Seconds since case start
                        x[trace_dim_index][padding+event_index][offset] = (event_time.timestamp() - case_start_time)/self.time_scaling_divisor[0]
                        # Seconds since previous event
                        x[trace_dim_index][padding+event_index][offset + 1] = (event_time.timestamp() - previous_event_time)/self.time_scaling_divisor[1]
                        # Encode additional time attributes if option is chosen
                        if self.advanced_time_attributes:
                            # Seconds since midnight
                            x[trace_dim_index][padding+event_index][offset + 2] = (event_time.hour * 3600 + event_time.second)/self.time_scaling_divisor[2]
                            # Seconds since last Monday
                            x[trace_dim_index][padding+event_index][offset + 3] = (event_time.weekday() * 86400 + event_time.hour * 3600 + event_time.second)/self.time_scaling_divisor[3]
                            # Seconds since last Januar 1
                            x[trace_dim_index][padding+event_index][offset + 4] = ((event_time.month - 1) * 2628000 + (event_time.day - 1) * 86400 + event_time.hour * 3600 + event_time.second)/self.time_scaling_divisor[4]
                            # Seconds since process start
                            x[trace_dim_index][padding+event_index][offset + 5] = (event_time.timestamp() - self.process_start_time)/self.time_scaling_divisor[5]

                            offset += 6
                        else:
                            offset += 2

                        previous_event_time = event_time.timestamp()

                        # Encode categorical attributes
                        for attribute_index, attribute in enumerate(self.categorical_attributes):
                            if event[attribute] in self.categorical_attributes_values[attribute_index]:
                                x[trace_dim_index][padding+event_index][offset + self.categorical_attributes_values[attribute_index].index(event[attribute])] = 1
                            offset += len(self.categorical_attributes_values[attribute_index])

                        # Encode numerical attributes
                        for attribute_index, attribute in enumerate(self.numerical_attributes):
                            x[trace_dim_index][padding+event_index][offset] = float(event[attribute])/self.numerical_divisor[attribute_index]
                            offset += 1

                        # Encode textual attribute
                        if self.text_encoder is not None and self.text_attribute is not None and self.text_attribute in event:
                            text_vectors = self.text_encoder.transform([event[self.text_attribute]])
                            x[trace_dim_index][padding+event_index][offset:offset+len(text_vectors[0])] = text_vectors[0]
                            offset += self.text_encoder.encoding_length

                # Set activity and time (since case start) of next event as target
                if for_training:
                    if prefix_length == len(case):
                        # Case 1: Set <Process end> as next activity target
                        y_next_act[trace_dim_index][len(self.activities)] = 1
                        y_next_time[trace_dim_index] = 0
                    else:
                        # Case 2: Set next activity as target
                        y_next_act[trace_dim_index][self.activities.index(case[prefix_length]["concept:name"])] = 1
                        y_next_time[trace_dim_index] = (case[prefix_length]["time:timestamp"].timestamp() - case[prefix_length - 1]["time:timestamp"].timestamp()) / self.time_scaling_divisor[1]
                    # Set final activity and case cycle time as target
                    y_final_act[trace_dim_index][self.activities.index(case[-1]["concept:name"])] = 1
                    y_final_time[trace_dim_index] = (case[-1]["time:timestamp"].timestamp() - case_start_time) / self.time_scaling_divisor[0]

                # Increase index for next (prefix-)trace
                trace_dim_index += 1

        if for_training:
            return x, y_next_act, y_final_act, y_next_time, y_final_time
        else:
            return x


def _get_event_labels(log, attribute_name):
    return list(dict.fromkeys([event[attribute_name] for case in log for event in case])) if log else []


def _is_numerical(value):
    try:
        float(value)
        return True
    except ValueError:
        return False


def _is_numerical_attribute(log, attribute):
    return _is_numerical(log[0][0][attribute])


def _get_max_case_length(log):
    return max([len(case) for case in log]) if log else 0


tapp/text_encoder.py
import numpy as np
import nltk
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from sklearn import utils
from sklearn.feature_extraction.text import TfidfVectorizer
from gensim.models import Doc2Vec
from gensim.models.doc2vec import TaggedDocument
from gensim.models import LdaModel
from gensim.corpora.dictionary import Dictionary
from abc import ABC, abstractmethod


class TextEncoder(ABC):

    def __init__(self, language="english", encoding_length=50):
        self.language = language

        try:
            nltk.data.find("corpora/wordnet")
        except LookupError:
            nltk.download("wordnet")
        try:
            nltk.data.find("tokenizers/punkt")
        except LookupError:
            nltk.download("punkt")
        try:
            nltk.data.find("corpora/stopwords")
        except LookupError:
            nltk.download("stopwords")

        self.encoding_length = encoding_length
        self.stop_words = set(stopwords.words(language))
        self.lemmatizer = WordNetLemmatizer() if language == "english" else None
        self.stemmer = SnowballStemmer(language)

        super().__init__()

    def preprocess_docs(self, docs, as_list=True):
        docs_preprocessed = []
        for doc in docs:
            doc = doc.lower()
            words = word_tokenize(doc, language=self.language)
            if self.lemmatizer is not None:
                words_processed = [self.lemmatizer.lemmatize(word) for word in words if word not in self.stop_words and word.isalpha()]
            else:
                words_processed = [self.stemmer.stem(word) for word in words if word not in self.stop_words and word.isalpha()]
            if as_list:
                docs_preprocessed.append(words_processed)
            else:
                docs_preprocessed.append(" ".join(words_processed))
        return docs_preprocessed

    @abstractmethod
    def fit(self, docs):
        pass

    @abstractmethod
    def transform(self, docs):
        pass


class BoWTextEncoder(TextEncoder):

    def __init__(self, language="english", encoding_length=100):
        self.name = "BoW"
        self.vectorizer = None
        super().__init__(language=language, encoding_length=encoding_length)

    def fit(self, docs):
        self.vectorizer = TfidfVectorizer(ngram_range=(1, 1), max_features=self.encoding_length, analyzer='word', norm="l2")
        self.vectorizer.fit(self.preprocess_docs(docs, as_list=False))
        return self

    def transform(self, docs):
        return self.vectorizer.transform(self.preprocess_docs(docs, as_list=False)).toarray()


class BoNGTextEncoder(TextEncoder):

    def __init__(self, language="english", encoding_length=100, n=2):
        self.name = "BoNG"
        self.n = n
        self.vectorizer = None
        super().__init__(language=language, encoding_length=encoding_length)

    def fit(self, docs):
        self.vectorizer = TfidfVectorizer(ngram_range=(self.n, self.n), max_features=self.encoding_length, analyzer='word', norm="l2")
        self.vectorizer.fit(self.preprocess_docs(docs, as_list=False))
        return self

    def transform(self, docs):
        return self.vectorizer.transform(self.preprocess_docs(docs, as_list=False)).toarray()


class PVTextEncoder(TextEncoder):

    def __init__(self,  language="english", encoding_length=20, epochs=15, min_count=2):
        self.name = "PV"
        self.epochs = epochs
        self.min_count = min_count
        self.model = None
        super().__init__(language=language, encoding_length=encoding_length)

    def fit(self, docs):
        docs = self.preprocess_docs(docs)

        tagged_docs = [TaggedDocument(words=doc, tags=[i]) for i, doc in enumerate(docs)]

        self.model = Doc2Vec(dm=1, vector_size=self.encoding_length, min_count=self.min_count, window=8)
        self.model.build_vocab(tagged_docs)

        self.model.train(utils.shuffle(tagged_docs), total_examples=len(tagged_docs), epochs=self.epochs)
        return self

    def transform(self, docs):
        docs = self.preprocess_docs(docs)
        return np.array([self.model.infer_vector(doc) for doc in docs])


class LDATextEncoder(TextEncoder):

    def __init__(self, language="english", encoding_length=20):
        self.name = "LDA"
        self.model = None
        self.num_topics = encoding_length
        self.dictionary = None
        super().__init__(language=language, encoding_length=encoding_length)

    def fit(self, docs):
        docs = self.preprocess_docs(docs)
        self.dictionary = Dictionary(docs)
        corpus = [self.dictionary.doc2bow(doc) for doc in docs]
        self.model = LdaModel(corpus, id2word=self.dictionary, num_topics=self.num_topics, minimum_probability=0.0)
        return self

    def transform(self, docs):
        docs = self.preprocess_docs(docs)
        docs = [self.dictionary.doc2bow(doc) for doc in docs]
        return np.array([self.model[doc] for doc in docs])[:, :, 1]


tapp/tapp_model.py
import numpy as np
import pandas as pd
import tensorflow as tf
from datetime import datetime
import os
from tensorflow.keras import metrics
from tensorflow.keras import layers
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from tensorflow.keras.losses import mean_absolute_error
from .prediction_model import PredictionModel


class TappModel(PredictionModel):

    def __init__(self, log_encoder=None, num_shared_layer=1, num_specialized_layer=1, neurons_per_layer=100, dropout=0.2, learning_rate=0.002):
        self.log_encoder = log_encoder
        self.num_shared_layer = num_shared_layer
        self.num_specialized_layer = num_specialized_layer
        self.neurons_per_layer = neurons_per_layer
        self.learning_rate = learning_rate
        self.dropout = dropout
        self.activities = []
        self.timesteps = 0
        self.feature_dim = 0
        self.model = None
        super().__init__()

    def _build_model(self):
        # Input layer
        inputs = tf.keras.Input(shape=(self.timesteps, self.feature_dim), name="inputs")

        # Shared layer
        previous_layer = inputs
        for layer in range(self.num_shared_layer):
            # Do not return sequences in the last LSTM layer
            return_sequences = False if self.num_specialized_layer == 0 and layer == self.num_shared_layer - 1 else True
            shared_lstm_layer = layers.LSTM(self.neurons_per_layer, implementation=2, kernel_initializer="glorot_uniform", return_sequences=return_sequences, dropout=self.dropout)(previous_layer)
            shared_lstm_normalization_layer = layers.LayerNormalization()(shared_lstm_layer)
            previous_layer = shared_lstm_normalization_layer

        previous_layer_next_activity = previous_layer
        previous_layer_final_activity = previous_layer
        previous_layer_next_timestamp = previous_layer
        previous_layer_final_timestamp = previous_layer
        for layer in range(self.num_specialized_layer):
            # Do not return sequences in the last LSTM layer
            return_sequences = False if layer == self.num_specialized_layer - 1 else True
            next_activity_lstm_layer = layers.LSTM(self.neurons_per_layer, implementation=2, kernel_initializer="glorot_uniform", return_sequences=return_sequences, dropout=self.dropout)(previous_layer_next_activity)
            final_activity_lstm_layer = layers.LSTM(self.neurons_per_layer, implementation=2, kernel_initializer="glorot_uniform", return_sequences=return_sequences, dropout=self.dropout)(previous_layer_final_activity)
            next_timestamp_lstm_layer = layers.LSTM(self.neurons_per_layer, implementation=2, kernel_initializer="glorot_uniform", return_sequences=return_sequences, dropout=self.dropout)(previous_layer_next_timestamp)
            final_timestamp_lstm_layer = layers.LSTM(self.neurons_per_layer, implementation=2, kernel_initializer="glorot_uniform", return_sequences=return_sequences, dropout=self.dropout)(previous_layer_final_timestamp)
            next_activity_normalization_layer = layers.LayerNormalization()(next_activity_lstm_layer)
            final_activity_normalization_layer = layers.LayerNormalization()(final_activity_lstm_layer)
            next_timestamp_normalization_layer = layers.LayerNormalization()(next_timestamp_lstm_layer)
            final_timestamp_normalization_layer = layers.LayerNormalization()(final_timestamp_lstm_layer)
            previous_layer_next_activity = next_activity_normalization_layer
            previous_layer_final_activity = final_activity_normalization_layer
            previous_layer_next_timestamp = next_timestamp_normalization_layer
            previous_layer_final_timestamp = final_timestamp_normalization_layer

        # Output layer
        next_activity_output = layers.Dense(len(self.activities) + 1, activation="softmax", kernel_initializer="glorot_uniform", name="next_activity_output")(previous_layer_next_activity)
        final_activity_output = layers.Dense(len(self.activities), activation="softmax", kernel_initializer="glorot_uniform", name="final_activity_output")(previous_layer_final_activity)
        next_timestamp_ouput = layers.Dense(1, kernel_initializer="glorot_uniform", name="next_timestamp_output")(previous_layer_next_timestamp)
        final_timestamp_ouput = layers.Dense(1, kernel_initializer="glorot_uniform", name="final_timestamp_output")(previous_layer_final_timestamp)

        # Build and configure model
        model = tf.keras.Model(inputs=[inputs], outputs=[next_activity_output, final_activity_output, next_timestamp_ouput, final_timestamp_ouput])
        optimizer_param = tf.keras.optimizers.Nadam(learning_rate=self.learning_rate, beta_1=0.9, beta_2=0.999, epsilon=1e-07, name="Nadam")
        metric_param = {"next_activity_output": metrics.CategoricalAccuracy(), "final_activity_output": metrics.CategoricalAccuracy(), "next_timestamp_output": metrics.MeanAbsoluteError(), "final_timestamp_output": metrics.MeanAbsoluteError()}
        loss_param = {"next_activity_output": "categorical_crossentropy", "final_activity_output": "categorical_crossentropy", "next_timestamp_output": "mae", "final_timestamp_output": "mae"}
        model.compile(loss=loss_param, metrics=metric_param, optimizer=optimizer_param)
        self.model = model

    def fit(self, log, data_attributes=None, text_attribute=None, epochs=100, validation_split=0.2):

        # Encode training data
        self.activities = _get_event_labels(log, "concept:name")
        self.log_encoder.fit(log, activities=self.activities, data_attributes=data_attributes, text_attribute=text_attribute)
        x, y_next_act, y_final_act, y_next_time, y_final_time = self.log_encoder.transform(log, for_training=True)

        # Build model
        self.timesteps = x.shape[1]
        self.feature_dim = x.shape[2]
        self._build_model()

        # Reduce learning rate if metrics do not improve anymore
        reduce_lr = ReduceLROnPlateau(monitor="val_loss", factor=0.1, patience=5, min_lr=0.0001)
        # Stop early if metrics do not improve for longer time
        early_stopping = EarlyStopping(monitor="val_loss", patience=10)
        # Fit the model to data
        return self.model.fit(x, {"next_activity_output": y_next_act, "final_activity_output": y_final_act, "next_timestamp_output": y_next_time, "final_timestamp_output": y_final_time}, epochs=epochs, batch_size=None, validation_split=validation_split, callbacks=[reduce_lr, early_stopping])

    def predict_next_activity(self, log):
        x = self.log_encoder.transform(log, for_training=False)
        prediction = self.model.predict(x)
        return prediction[0]

    def predict_outcome(self, log):
        x = self.log_encoder.transform(log, for_training=False)
        prediction = self.model.predict(x)
        return prediction[1]

    def predict_next_time(self, log):
        x = self.log_encoder.transform(log, for_training=False)
        prediction = self.model.predict(x)
        return np.maximum(prediction[2].flatten() * self.log_encoder.time_scaling_divisor[1], 0)

    def predict_cycle_time(self, log):
        x = self.log_encoder.transform(log, for_training=False)
        prediction = self.model.predict(x)
        return np.maximum(prediction[3].flatten() * self.log_encoder.time_scaling_divisor[0], 0)

    def predict_suffix(self, log):
        return

    def predict(self, log):
        x = self.log_encoder.transform(log, for_training=False)
        prediction = self.model.predict(x)
        prediction[2] = np.maximum(prediction[2].flatten() * self.log_encoder.time_scaling_divisor[1], 0)
        prediction[3] = np.maximum(prediction[3].flatten() * self.log_encoder.time_scaling_divisor[0], 0)
        return prediction

    def _evaluate_raw(self, log):
        # Make predictions
        prefix_log = [case[0:prefix_length] for case in log for prefix_length in range(1, len(case) + 1)]
        predictions = self.predict(prefix_log)
        predicted_next_activities = np.argmax(predictions[0], axis=1)
        predicted_case_outcomes = np.argmax(predictions[1], axis=1)
        predicted_next_times = predictions[2] / 86400
        predicted_cycle_times = predictions[3] / 86400
        caseIDs = []
        prefix_lengths = []
        true_next_activities = []
        true_case_outcomes = []
        true_next_times = []
        true_cycle_times = []
        for case in log:
            caseID = case.attributes["concept:name"]
            for prefix_length in range(1, len(case) + 1):
                caseIDs.append(caseID)
                prefix_lengths.append(prefix_length)

                true_next_activities.append(len(self.activities) if prefix_length == len(case) else self.activities.index(case[prefix_length]["concept:name"]) if case[prefix_length]["concept:name"] in self.activities else -1)
                true_case_outcomes.append(self.activities.index(case[-1]["concept:name"]) if case[-1]["concept:name"] in self.activities else -1)
                true_next_times.append(0 if prefix_length == len(case) else (case[prefix_length]["time:timestamp"].timestamp() - case[prefix_length - 1]["time:timestamp"].timestamp()) / 86400)
                true_cycle_times.append((case[-1]["time:timestamp"].timestamp() - case[0]["time:timestamp"].timestamp()) / 86400)

        # Generate DataFrame
        column_data = {"caseID": caseIDs, "prefix-length": prefix_lengths, "true-next-activity": true_next_activities, "pred-next-activity": predicted_next_activities, "true-outcome": true_case_outcomes, "pred-outcome": predicted_case_outcomes, "true-next-time": true_next_times, "pred-next-time": predicted_next_times, "true-cycle-time": true_cycle_times, "pred-cylce-time": predicted_cycle_times}
        columns = ["caseID", "prefix-length", "true-next-activity", "pred-next-activity", "true-next-time", "pred-next-time", "true-outcome", "pred-outcome", "true-cycle-time", "pred-cylce-time"]
        return pd.DataFrame(column_data, columns=columns)

    def evaluate(self, log, path, num_prefixes=8):
        # Generate raw predictions
        raw = self._evaluate_raw(log)
        # Compute metrics
        next_activity_acc = len(raw[(raw["pred-next-activity"] == raw["true-next-activity"]) & (raw["prefix-length"] > 1)]) / np.max([len(raw[raw["prefix-length"] > 1]), 1])
        next_time_mae = mean_absolute_error(raw[raw["prefix-length"] > 1]["true-next-time"].astype(float).to_numpy(), raw[raw["prefix-length"] > 1]["pred-next-time"].astype(float).to_numpy()).numpy()
        outcome_acc = len(raw[(raw["pred-outcome"] == raw["true-outcome"]) & (raw["prefix-length"] > 1)]) / np.max([len(raw[raw["prefix-length"] > 1]), 1])
        cycle_time_mae = mean_absolute_error(raw[raw["prefix-length"] > 1]["true-cycle-time"].astype(float).to_numpy(), raw[raw["prefix-length"] > 1]["pred-cylce-time"].astype(float).to_numpy()).numpy()

        next_activity_acc_pre = [len(raw[(raw["pred-next-activity"] == raw["true-next-activity"]) & (raw["prefix-length"] == prefix_length)]) / np.max([len(raw[raw["prefix-length"] == prefix_length]), 1]) for prefix_length in range(1, num_prefixes + 1)]
        next_time_mae_pre = [mean_absolute_error(raw[raw["prefix-length"] == prefix_length]["true-next-time"].astype(float).to_numpy(), raw[raw["prefix-length"] == prefix_length]["pred-next-time"].astype(float).to_numpy()).numpy() for prefix_length in range(1, num_prefixes + 1)]
        outcome_acc_pre = [len(raw[(raw["pred-outcome"] == raw["true-outcome"]) & (raw["prefix-length"] == prefix_length)]) / np.max([len(raw[raw["prefix-length"] == prefix_length]), 1]) for prefix_length in range(1, num_prefixes + 1)]
        cycle_time_mae_pre = [mean_absolute_error(raw[raw["prefix-length"] == prefix_length]["true-cycle-time"].astype(float).to_numpy(), raw[raw["prefix-length"] == prefix_length]["pred-cylce-time"].astype(float).to_numpy()).numpy() for prefix_length in range(1, num_prefixes + 1)]

        prefix_predictions = next_activity_acc_pre + next_time_mae_pre + outcome_acc_pre + cycle_time_mae_pre

        if not os.path.exists(path):
            prefix_columns = []
            for metric in ["naa_{}", "ntm_{}", "oa_{}", "ctm_{}"]:
                for prefix in range(1, num_prefixes + 1):
                    prefix_columns.append(metric.format(prefix))
            columns = ["model", "timestamp", "num_layer", "num_shared_layer", "hidden_neurons", "advanced_time_attributes", "data_attributes", "event_dim", "text_encoding", "text_dim", "next_activity_acc", "next_time_mae", "outcome_acc", "cycle_time_mae"] + prefix_columns
            df = pd.DataFrame(columns=columns)
            df.to_csv(path, encoding="utf-8", sep=",", index=False)
        df = pd.read_csv(path, sep=",")

        df.loc[len(df)] = ["tapp", datetime.now().strftime("%Y-%m-%d-%H-%M-%S"),self.num_shared_layer + self.num_specialized_layer, self.num_shared_layer, self.neurons_per_layer, self.log_encoder.advanced_time_attributes, self.log_encoder.data_attributes, self.log_encoder.feature_dim, self.log_encoder.text_encoder.name if self.log_encoder.text_encoder is not None else "-",
            self.log_encoder.text_encoder.encoding_length if self.log_encoder.text_encoder is not None else 0, next_activity_acc, next_time_mae, outcome_acc, cycle_time_mae] + prefix_predictions
        df.to_csv(path, encoding="utf-8", sep=",", index=False)
        return df

    def _get_activity_label(self, index):
        return self.activities[index]


def _get_event_labels(log, attribute_name):
    return list(dict.fromkeys([event[attribute_name] for case in log for event in case])) if log else []
